Simple Grover quickstart (#549)
MicrosoftDocs/quantum-docs-pr: articles/quickstarts/search.md

---
title: "Q# quickstart: TODO"
description: TODO
author: cgranade
ms.author: chgranad@microsoft.com
ms.date: 10/9/2019
ms.topic: article
uid: microsoft.quantum.quickstarts.search
---

# Search for marked inputs with Grover's Algorithm

This introduction uses [Grover's Algorithm][grover] to search inputs to a function for a particular _marked input_.

> This algorithm can be used to search for inputs with some special property, to find items in an unstructured list, and so forth.

Reviewer: ```suggestion
Grover's algorithm is used to search for inputs with some special property, to find items in an unstructured list.
```

Grover's algorithm lets us use a quantum computer to run this search in less steps than the number of items in the list that we're searching — something no classical algorithm can do.

> In this introduction, we see how we can use Q# and the Quantum Development Kit to run Grover's algorithm with just a few lines of code.

Reviewer: ```suggestion
You can use Q# and the Quantum Development Kit to run Grover's algorithm with just a few lines of code.
```

> ## Prerequisites

Reviewer: We should also spell out what a user will learn from this quickstart, or what result they will get. There is a standard list format for this. I can help you out with it

Author: I'd be much obliged for your help on that, thank you!

- The Microsoft [Quantum Development Kit][install].

> ## Writing Grover's Algorithm

Reviewer: It would be good to spell out the scenario before jumping into the code i.e. give the type of data input and what the algorithm will do with it. We could even have a diagram.

Using the Quantum Development Kit, create a new Q# project:

```

> $ dotnet new console -lang Q# --output Grover

Reviewer 2: Are we going to add instructions for how to do this in our other environments in tabs as well?

Author: I'm a bit worried about proliferating how many different copies of the project creation instructions we have kicking about, I guess.

Reviewer: We (I :-)) could create a very short "how-to" guide called "How to create a Q# project", and we could link to it, as a pre-requisite ...

$ cd Grover
```

Grover's algorithm itself can be written using just a few lines.
Add the following code to the `Operations.qs` file in your new project:

> [!code-qsharp[](~/quantum/samples/algorithms/simple-grover/SimpleGrover.qs?highlight=5,27)]

Reviewer 2: Generally for this kind of tutorial, does additional information live in code comments or as docs text with a snippet underneath? In the TensorFlow example the code is more broken up into steps rather than one large block to copy paste. Having the big block is easy for copy paste but doesn't really take advantage of it being a walk through (because you can copy it and run it without reading any of the comments). @natke , is there a docs standard for how to do this?

Author: I hope that this sample can get a bit shorter as we bring more of the contents of `Reflections.qs` into the libraries themselves as convenience operations.

Author: I definitely agree with breaking the code up in cases like the Bell state preparation sample, but my hope is that we can move a bunch of operations over to the M.Q.Preparation and M.Q.AmplitudeAmplification namespaces (e.g.: ReflectAboutUniform isn't really specific to this sample). As we do so, this article should automatically update thanks to the source code links that @natke helped to get working.

To define the list that we're searching, create a new file `Reflections.qs`, and paste in the following code:

> [!code-qsharp[](~/quantum/samples/algorithms/simple-grover/Reflections.qs)]

Reviewer 2: This code wasn't showing up for me, not sure if this is something that is being worked on or a new issue.

Author: As mentioned earlier, the code will not show up until microsoft/quantum#237 is completed. This PR is marked as a draft for that reason.

> The `ReflectAboutMarked` operation in this file represents the problem we're trying to solve with Grover's algorithm, in this case by reflecting bitstings made up of alternating zeros and ones.

Reviewer 2: typo: bitstrings ?

Author: Yep, that's a legit typo. Thanks for catching that.

> The `ReflectAboutMarked` operation in this file represents the problem we're trying to solve with Grover's algorithm, in this case by reflecting bitstings made up of alternating zeros and ones.

Reviewer 2: What purpose does reflecting bit strings serve? The goal of this tutorial is to demonstrate impact. Is it possible to use example input that shows a quantum advantage? Or add a couple sentences around why this is interesting? (I recognize that it's a delicate balance around the quantity or information, just wondering if a bit more detail makes sense)

Author: There's a couple sentences about that above, as well as a conclusion below that indicates that the right input is found in just four calls; would you recommend adding more detail here?

Reviewer 2: I think what I'm missing is the connection between specifically reflecting bit strings to the search algorithm. After re-reading, I think what might be good is to include before the conclusion some statement like "we are looking for 01010 in a list of n". I only really understood what we were searching the list for after reading the conclusion, so bringing that up would likely make it more clear and provide a nice path to the conclusion. Does that make sense?

Next, run your new Q# program to find the item marked by `ReflectAboutMarked`.

### [Python with Visual Studio Code or the Command Line](#tab/tabid-python)

```bash

> $ python host.py

Reviewer: Where do we define host.py?
